src/databricks/labs/lsql/dashboards.py

```diff
@@ -728,6 +766,118 @@ def _get_query_widget_spec(fields: list[Field], *, frame: WidgetFrameSpec | None
+
+    def validate(self) -> None:
+        """Validate the tile
+
+        Raises:
+            ValueError : If the tile is invalid.
+        """
+        if not self.metadata.is_filter():
+            raise ValueError(f"Tile is not a filter file: {self}")
+        if len(self.metadata.filters) == 0:
+            raise ValueError(f"Filter tile with path {self.metadata.path} has no filters defined")
+
+    def get_layouts(self, datasets: list[Dataset]) -> Iterable[Layout]:
+        """Get the layout(s) reflecting this tile in the dashboard."""
+        dataset_columns = self._get_dataset_columns(datasets)
+        if len(dataset_columns) == 0:
+            err_msg = f"Filter tile with path {self.metadata.path} has no matching dataset columns"
+            raise ValueError(err_msg)
+        widget = self._create_filter_widget(dataset_columns)
+        layout = Layout(widget=widget, position=self.position)
+        yield layout
+
+    def _create_filter_widget(self, dataset_columns: set[tuple[str, str]]) -> Widget:
+        filter_type = self.metadata.widget_type
+        match filter_type:
+            case WidgetType.MULTI_SELECT:
+                return self._create_multi_select_widget(dataset_columns)
+            case WidgetType.DATE_RANGE_PICKER:
+                return self._create_date_range_picker_widget(dataset_columns)
+            case _:
+                raise NotImplementedError(f"Filter type {filter_type} not supported for filter tile: {self}")
+
+    def _create_multi_select_widget(self, dataset_columns: set[tuple[str, str]]) -> Widget:
+        control_encodings: list[ControlEncoding] = []
+        queries = []
+        for idx, (column, dataset_name) in enumerate(dataset_columns):
+            filter_fields = [
+                Field(name=column, expression=f"`{column}`"),
+                Field(name=f"{column}_associativity", expression="COUNT_IF(`associative_filter_predicate_group`)"),
+            ]
+            filter_query = Query(dataset_name=dataset_name, fields=filter_fields, disaggregated=False)
+            filter_named_query = NamedQuery(name=f"multi_select_{column}_{idx}", query=filter_query)
+            queries.append(filter_named_query)
+            control_encodings.append(ControlFieldEncoding(column, filter_named_query.name, display_name=column))
+
+        frame = WidgetFrameSpec(
+            title=self.metadata.title,
+            show_title=len(self.metadata.title) > 0,
+            description=self.metadata.description,
+            show_description=len(self.metadata.description) > 0,
+        )
+        control_encoding_map = ControlEncodingMap(control_encodings)
+        spec = MultiSelectSpec(encodings=control_encoding_map, frame=frame)
+        widget = Widget(name=f"{self.metadata.id}_widget", queries=queries, spec=spec)
+        return widget
+
+    def _create_date_range_picker_widget(self, dataset_columns: set[tuple[str, str]]) -> Widget:
```

Comment:
Looks like largely a duplicate of the other method

Reply:
Although the code looks similar, it may evolve differently with future filter types and filter specific changes. Premature abstraction can complicate maintenance and reduce flexibility while changing individual widget types. It's better to wait for patterns to emerge automatically and create abstractions later based on those common patterns when all filter widgets are added. As they say, duplication is often cheaper than incorrect abstraction. I lean towards KISS than DRY.

Reply:
hmm, 'ke, I kinda get that. 

Given that this is an exact duplicate besides the spec name in the named query and the spec itself, I would have made this one method myself.

Note that you implemented feedback I gave earlier only on one of the methods, missing to implement the other where the same feedback applied because it is a duplicate

Reply:
Yes, that's what was intended in the comments. I don't see a contradiction here about duplication. The point was to wait till we add the other filter widgets and then separate the common parts as we will know better what's common and what's not across all the filters we intend to support. Will separate out the common parts assuming Multiselect, DateRange and Dropdown are the only Filter widgets we will support for now.